substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/ThreadMXUtils.java

```diff
@@ -0,0 +1,206 @@
+            for (int i = 0; i < lockObjects.length; i++) {
+                lockObjects[i] = locks.iterator().next();
+            }
+            return lockObjects;
+        }
+
+        private record LockedMonitors(Object[] monitorObjects, int[] monitorDepths) {
+        }
+
+        private static LockedMonitors getLockedMonitors(Thread thread) {
+            EconomicSet<JavaMonitor> monitors = JMXMonitoring.getThreadMonitors(thread);
+            Object[] monitorObjects = new Object[monitors != null ? monitors.size() : 0];
+            int[] monitorDepths = new int[monitorObjects.length];
+            for (int i = 0; i < monitorObjects.length; i++) {
+                JavaMonitor javaMonitor = monitors.iterator().next();
+                monitorObjects[i] = javaMonitor.getBlockedObject();
+                monitorDepths[i] = javaMonitor.getDepth();
+            }
+            return new LockedMonitors(monitorObjects, monitorDepths);
+        }
+
+        public static ThreadInfo getThreadInfo(Thread thread, int maxDepth,
+                        boolean withLockedMonitors, boolean withLockedSynchronizers) {
+            assert thread != null;
+            Blocker blocker = getBlockerInfo(thread);
+            LockedMonitors lockedMonitors = getLockedMonitors(thread);
+
+            Target_java_lang_management_ThreadInfo targetThreadInfo = new Target_java_lang_management_ThreadInfo(
+                            thread,
+                            getThreadState(thread),
+                            blocker.blockObject,
+                            blocker.ownerThread,
+                            JMXMonitoring.getThreadTotalBlockedCount(thread),
+                            JMXMonitoring.getThreadTotalBlockedTime(thread),
+                            JMXMonitoring.getThreadTotalWaitedCount(thread),
+                            JMXMonitoring.getThreadTotalWaitedTime(thread),
+                            getStackTrace(thread, maxDepth),
+                            withLockedMonitors ? lockedMonitors.monitorObjects : new Object[0],
+                            withLockedMonitors ? lockedMonitors.monitorDepths : new int[0],
+                            withLockedSynchronizers ? getLockedSynchronizers(thread) : new Object[0]);
+            return fromTarget(targetThreadInfo);
+        }
+
+        @SuppressFBWarnings(value = "BC", justification = "Cast for @TargetClass")
+        private static ThreadInfo fromTarget(Target_java_lang_management_ThreadInfo targetThreadInfo) {
+            return ThreadInfo.class.cast(targetThreadInfo);
+        }
+    }
+
+    public static class DeadlockDetectionUtils {
+        private static final ReentrantLock lock = new ReentrantLock();
+        private static final HashSet<Long> deadlocked = new HashSet<>();
+        private static final HashSet<Long> chain = new HashSet<>();
+        private static ThreadInfo[] allThreadInfos = new ThreadInfo[0];
```

Comment:
Is this global state really needed? Could you allocate the objects in `findDeadlockedThreads` and pass them to the other private helpers instead?